### ISSUE TITLE
Consider making Iterator.empty's type consistent with Seq.empty

### ISSUE THREAD

**Issue by klaeufer** [scala/bug#10711]
# Version info

Scala 2.12.4

# Description

Currently, one can easily create an empty `Seq` of a specific item type

```
Seq.empty[Int]
```

but not an empty `Iterator`:

```
(Iterator.empty: Iterator[Int])
Seq.empty[Int].iterator
```

This ability is important for left-to-right typing of expressions involving higher-order functions on iterators, such as

```
Iterator.empty[Int].scanLeft(0)(_ + _)
```

# Proposed fix

Change 

https://github.com/scala/scala/blob/v2.12.4/src/library/scala/collection/Iterator.scala#L36

to

```
    def empty[T]: Iterator[T] = new AbstractIterator[Nothing] {
```

**Comment by viktorklang** [comment]
@klaeufer Allocating a new Iterator for those invocations should be avoided, i.e. I want to preserve current behavior:

~~~
Welcome to Scala 2.12.3 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_144).
Type in expressions for evaluation. Or try :help.

scala> Iterator.empty eq Iterator.empty
res0: Boolean = true
~~~

So I propose: 

~~~scala
  private[this] final val _empty: Iterator[Nothing] = new AbstractIterator[Nothing] {
    override def hasNext: Boolean = false
    override def next(): Nothing = throw new NoSuchElementException("next on empty iterator")
  }
  def empty[T]: Iterator[T] = _empty
~~~

**Comment by klaeufer** [comment]
@viktorklang That's a much better solution!

**Comment by joshlemer** [comment]
@klaeufer this is fixed as of 2.13, this issue can be closed now.
